images/cfw/opt/x1plus/lib/python/x1plus/services/x1plusd/__init__.py

```diff
@@ -15,44 +20,192 @@
 from .actions import ActionHandler
 from .gpios import GpioManager
 
+from x1plus.utils import module_loader, module_docstring_parser
+
 logger = logging.getLogger(__name__)
 
 class X1PlusDaemon:
+
+    MODULES = {}    
+    """
+    MODULES = {
+        "name": {
+            "definition": {},
+            "module_class": Class,
+            "loaded": true/false. False if error loading.
+        }
+    } 
+    """
+
+
     @classmethod
     async def create(cls):
         # must be a classmethod, since __init__ cannot be async
         self = X1PlusDaemon()
 
         logger.info("creating x1plusd services")
+        
+        # Required modules
         self.router = await get_dbus_router()
         self.settings = SettingsService(router=self.router)
+
         self.mqtt = MQTTClient(daemon=self)
         self.ota = OTAService(router=self.router, daemon=self)
         self.ssh = SSHService(daemon=self)
         self.httpd = HTTPService(router=self.router, daemon=self)
         self.sensors = SensorsService(router=self.router, daemon=self)
-        self.gpios = GpioManager(daemon=self)
-        self.expansion = ExpansionManager(router=self.router, daemon=self)
         self.mcproto = MCProtoParser(daemon=self)
+        self.expansion = ExpansionManager(router=self.router, daemon=self)
+        self.gpios = GpioManager(daemon=self)
         self.actions = ActionHandler(router=self.router, daemon=self)
-        if not self.settings.get("polar_cloud", False):
-            self.polar_cloud = None
-        else:
-            from .polar_cloud import PolarPrintService
-            self.polar_cloud = PolarPrintService(daemon=self)
         
+        self.custom_modules = {}
+        """
+        { "name": ClassInstance }
+        """
+
+        self.settings_keys = []
+
+        BASE_MODULE_DIR = os.path.dirname(os.path.abspath(__file__))
+        # Custom module folder could be in x1plus.services.x1plusd.custom, and symlinked to sd card?
+
+        directories = [BASE_MODULE_DIR]
+
+        for filename in os.listdir(BASE_MODULE_DIR):
+            full_path = os.path.join(BASE_MODULE_DIR, filename)
+            if os.path.isdir(full_path) and not filename.startswith("_"):
+                directories.append(full_path)
```

Comment:
I did build this initially for there being able to work with any directory, and much prefer a "here's a single directory with everything" approach. That would just require some architectural changes for say, polar_cloud, which I think given from other comments, is probably fine to do now. 

Subdirs could be used for actions/drivers/animations, more so of a means of organization.

As for the sd card path, that can easily be added as a second directory to look into, or we could symlink the base module directory to the sd card. Either one works I think.

Reply:
one other thing I was thinking about in the shower a few days ago was the concept of distributing multiple python files together as a module.  I did not come up with any good solutions, and I don't think it has to go into this version -- single file load would still be a huge improvement.  but here are some things I was thinking about as possible design constraints for improvements later:

* the general idea is that it would be sort of annoying that current Modules have to be exactly one file -- or otherwise you have to keep track of them separately.  for instance, if the whole expansion subsystem were an X1Plus Module, you would be copying many files to the SD card, and versioning them all independently, but presumably a plugin developer would want to release one file that gets loaded in atomically.
* one thing that we could do is load Python eggs, which are just zip files that can be stuck in a pythonpath, as I understand it.  seems ok, but eggs are deprecated, and more to the point, if there need to be native components with native binaries, then they can't really go into eggs.  and even more likely, it might be the case that at some point we have a mechanism to load qml files into bbl_screen -- and those would want to have qml files in the same directory as py files.  that would not work well with eggs either.
* I wonder, then, if in the long term, X1Plus Modules are something that get installed to /userdata (or to the rootfs overlay, and then wiped out when X1Plus is reinstalled?), and are just zipfiles that the front-end unpacks?

anyway, these are things to think about.  I don't like any of these answers yet, and since I don't like anything particularly better than what you propose so far, I'm definitely ok with continuing down this path as at least one option!  and anyway, single-file python Modules should always be supported, since they are a lightweight way to get stuff running.